### PR TITLE
add dismissible notification when student scores are available

### PR DESCRIPTION
**Pull request by davidsbailey**
Finishes implementation work for [AITT-816: Alert teachers of waiting scores - US 3](https://codedotorg.atlassian.net/browse/AITT-816). all that remains is some fit-and-finish changes and to launch the experiment.

the requirement is to show the alert only once on each lesson. to implement this, I'm storing a map of lesson positions for each user. this makes it so that dismissing the alert in one lesson (e.g. lesson 11) in one of the many copies of csd3 (e.g. `interative-animations-games-2024`) will also hide the alert on the same lesson in other copies of the same unit (e.g. `csd3-2024`). 

The alerts are limited to csd-2023 and csd-2024 so that we do not unintentionally keep showing these alerts beyond the current version year -- if we decide to show these for 2025 that will be easy to change later.

## screenshots

each sequence starts with clearing the user's settings so that they have not "seen" the alert on any lessons.

### dismiss alert via close button

https://github.com/user-attachments/assets/7845ce07-e4e1-4a04-b86b-16b6af323b40

### view scores

https://github.com/user-attachments/assets/7adde5c8-bc22-4dbb-836a-eebb14969a3d

### clicking fab dismisses alert

https://github.com/user-attachments/assets/a9b83782-0dd9-4e01-b171-aeb826e8cf87

### auto dismiss alert if TA window is already open

https://github.com/user-attachments/assets/e47906e5-2bf6-4223-a072-f9484b9b0def

## Testing story
* manual testing as shown in screenshots
* js tests
* ruby integration tests instead of controller tests, which make it possible to test the interaction between routes from different controllers
* no green check for drone yet, but I do now have a passing [unit](https://drone.cdn-code.org/code-dot-org/code-dot-org/18997/1/6) and [UI test](https://drone.cdn-code.org/code-dot-org/code-dot-org/18961/2/6) run across multiple drone runs

### CODE REVIEW

File: dashboard/app/controllers/script_levels_controller.rb [hunk 1]
```diff
@@ -627,4 +628,12 @@ def self.get_script(request)
 
     redirect_script
   end
+
+  private def can_show_ta_scores_alert?
+    course_version = @script_level&.script&.get_course_version
+    return false unless course_version&.course_offering&.key == 'csd'
+    return false unless ['2023', '2024'].include?(course_version.key)
+    seen_ta_scores_map = current_user&.seen_ta_scores_map || {}
+    !seen_ta_scores_map[@script_level.lesson.absolute_position.to_s]
```
Comment (etaderhold) [inline dashboard/app/controllers/script_levels_controller.rb]: Is this `lesson.absolute_position` field on the backend a synonym for the `lesson.position` field that is used as a map key on the frontend?

**Review by etaderhold — COMMENTED** [review]
I have some questions about the "seen" notification and the design for the map storing when stuff was seen. It looks like it may work just fine for a limited beta use case but would likely need rework in the future if the TA functionality is to be used in more units and/or left in production for multiple school years.

1. For teachers with multiple sections, are the AI scores generated at the same time for all the sections? If not, might we want to notify a teacher about the scores being available for Section 2 even if they've already seen and dismissed the message about scores being available for Section 1?
2. How stable are lesson numbers from year to year, not just in this one CSD unit we currently support, but generally? Are lessons sometimes inserted or removed between version years, such that lesson 11 in 2023 might not be the same as lesson 11 in 2024? If so, would using lesson names as the key be a better idea?
3. What's the plan for expanding the teaching assistant beyond the one unit where it's currently available? If this is happening anytime soon, would it be wise to include more than just the lesson number as the key here, adding course and unit info as well (and maybe section)? Putting this in from the start would avoid the need to migrate the existing data to a new format later.
4. If the teacher teaches the same course again next semester would we want to show them the notification again when the next crop of students does their work and has scores available?
5. Does this "seen" information need to be shared across all of the user's machines and browsers, or would local storage be sufficient, and we could live with it if a teacher sees a repeat notification when they sign in from a different computer?

**Comment by davidsbailey** [comment]
> I have some questions about the "seen" notification and the design for the map storing when stuff was seen. It looks like it may work just fine for a limited beta use case but would likely need rework in the future if the TA functionality is to be used in more units and/or left in production for multiple school years.
> 
> 1. For teachers with multiple sections, are the AI scores generated at the same time for all the sections? If not, might we want to notify a teacher about the scores being available for Section 2 even if they've already seen and dismissed the message about scores being available for Section 1?
> 2. How stable are lesson numbers from year to year, not just in this one CSD unit we currently support, but generally? Are lessons sometimes inserted or removed between version years, such that lesson 11 in 2023 might not be the same as lesson 11 in 2024? If so, would using lesson names as the key be a better idea?
> 3. What's the plan for expanding the teaching assistant beyond the one unit where it's currently available? If this is happening anytime soon, would it be wise to include more than just the lesson number as the key here, adding course and unit info as well (and maybe section)? Putting this in from the start would avoid the need to migrate the existing data to a new format later.
> 4. If the teacher teaches the same course again next semester would we want to show them the notification again when the next crop of students does their work and has scores available?
> 5. Does this "seen" information need to be shared across all of the user's machines and browsers, or would local storage be sufficient, and we could live with it if a teacher sees a repeat notification when they sign in from a different computer?

Great questions Eric -- the first thing to know is that the feature request is a bit underspecified and says only to show the alert "once per lesson", which I've interpreted in the most conservative way possible so as to limit total exposures for each teacher to about 6. My interpretation is that the alert is to inform the user of what the 🔴 2️⃣  notification badge means and therefore shouldn't need to be shown for every section or for every "copy" of csd3 or every year or on every computer.

we should assume that lesson positions, names, and "keys" are not stable in general. We should also assume that more lessons will have TA enabled in 2025 and beyond.

I'm leaning toward (1) lobby for just showing the alert once per user, or (2) come up with a more robust way to limit the number of exposures to a reasonable number. 

for (2), perhaps for example we would show this alert on the first N lessons the teacher sees (tracked in the hashmap by lesson id rather than lesson position), and once there are N entries in the map we would never show the alert again. LMK how this sounds and perhaps we can discuss with Tess and Molly at checkin.

File: dashboard/app/controllers/script_levels_controller.rb (hunk 1 above)
Comment (davidsbailey) [reply dashboard/app/controllers/script_levels_controller.rb]: moot now, but good catch -- it looks like we were using relative_position for lesson.position, which can be different if there are any lockable lessons (exams/assessments) in the unit. https://github.com/code-dot-org/code-dot-org/blob/43814587fc56f14a799061f458a385cd1002ac1e/dashboard/app/models/rubric.rb#L35

**Review by etaderhold — APPROVED** [review]
Nice!

**Pull request merged**